src/dialogs/editor.tsx

```diff
@@ -0,0 +1,203 @@
+        } catch (exc: any /* eslint-disable-line @typescript-eslint/no-explicit-any */) {
+            this.update({ error: cockpit.message(exc), saving: false });
+        }
+    }
+
+    close() {
+        this.file.close();
+    }
+}
+
+export const EditFileModal = ({ dialogResult, path } : {
+    dialogResult: DialogResult<void>,
+    path: string
+}) => {
+    const [state, setState] = React.useState(new EditorState());
+    const [editor, setEditor] = React.useState<Editor | null>(null);
+
+    React.useEffect(() => {
+        const editor = new Editor(path);
+        editor.on('updated', setState);
+        setEditor(editor);
+        return () => {
+            editor.close();
+        };
+    }, [path]);
+
+    const modified = state.modified;
+    React.useEffect(() => {
+        const before_unload = (event: BeforeUnloadEvent) => {
+            event.preventDefault();
+        };
+        if (modified) {
+            window.addEventListener('beforeunload', before_unload);
+            return () => {
+                window.removeEventListener('beforeunload', before_unload);
+            };
+        }
+    }, [modified]);
+
+    /* Translators: $0 represents a filename */
+    const title = cockpit.format(state?.writable ? _("Editing “$0”") : _("Viewing “$0”"), path);
+
+    return (
+        <Modal
+          position="top"
+          title={(modified ? "⭐ " : "") + title}
+          isOpen
+          onClose={() => dialogResult.resolve()}
+          variant={ModalVariant.large}
+          className='file-editor-modal'
+          footer={
+              state?.writable &&
+              <>
+                  <Button
+                    variant="primary"
+                    isDisabled={
+                        !editor ||
+                            state.saving ||
+                            !modified ||
+                            !state.writable ||
+                            state.tag_now !== state.tag_at_load
+                    }
+                    onClick={() => editor && editor.save()}
+                  >
+                      {_("Save")}
+                  </Button>
+                  <Button variant="link" onClick={() => dialogResult.resolve()}>
+                      {modified ? _("Cancel (discard changes)") : _("Cancel")}
+                  </Button>
+              </>
+          }
+        >
+            <Stack>
+                {state.error !== null &&
+                <Alert
+                  type="danger"
+                  title={state.error}
+                  isInline
+                />}
+                {state.tag_now !== state.tag_at_load &&
+                <Alert
+                  isInline
+                  variant="warning"
+                  title="The file has changed on disk"
+                  actionLinks={
+                      <>
+                          <AlertActionLink onClick={() => editor && editor.load_file()}>
```

Comment:
The buttons shouldn't even appear (or be `isDisabled`) if the editor component is initialized asynchronously here. An enabled button should never do a no-op on click.

(Aside from that, in other situations this is better written as `editor?.load_file()`, but it doesn't apply here)